src/cluster.cc

```diff
@@ -140,6 +161,134 @@ void Cluster::SetMasterSlaveRepl() {
+    return Status(Status::NotOK, "Slave can't migrate slot");
+  }
+  if (nodes_[dst_node]->role_ != kClusterMaster) {
+    return Status(Status::NotOK, "Can't migrate slot to a slave");
+  }
+  if (nodes_[dst_node] == myself_) {
+    return Status(Status::NotOK, "Can't migrate slot to myself");
+  }
+
+  const auto dst = nodes_[dst_node];
+  Status s = svr_->slot_migrate_->MigrateStart(svr_, dst_node, dst->host_,
+                                               dst->port_, slot,
+                                               svr_->GetConfig()->migrate_speed,
+                                               svr_->GetConfig()->pipeline_size,
+                                               svr_->GetConfig()->sequence_gap);
+  return s;
+}
+
+Status Cluster::ImportSlot(Redis::Connection *conn, int slot, int state) {
+  if (myself_->role_ != kClusterMaster || svr_->IsSlave()) {
+    return Status(Status::NotOK, "Slave can't import slot");
+  }
+  if (slot < 0 || slot >= HASH_SLOTS_SIZE) {
+    return Status(Status::NotOK, "Slot is out of range");
+  }
+
+  switch (state) {
+    case kImportStart:
+      if (svr_->slot_import_->Start(conn->GetFD(), slot) == false) {
+        return Status(Status::NotOK, "Can't start importing slot " + std::to_string(slot));
+      } else {
+        // Set link importing
+        conn->SetImporting();
+        myself_->importing_slot_ = slot;
+        // Set link error callback
+        conn->close_cb_ = std::bind(&SlotImport::StopForLinkError, svr_->slot_import_, conn->GetFD());
+        // Stop forbiding writing slot to accept write commands
+        svr_->slot_migrate_->ReleaseForbiddenSlot();
+        LOG(INFO) << "[import] Start importing slot " << slot;
+      }
+      break;
+    case kImportSuccess:
+      if (svr_->slot_import_->Success(slot) == false) {
+        LOG(ERROR) << "[import] Failed to set slot importing success, maybe slot is wrong"
+                  << ", recieved slot: " << slot
+                  << ", current slot: " << svr_->slot_import_->GetSlot();
+        return Status(Status::NotOK, "Failed to set slot " + std::to_string(slot) + " importing success");
+      }
+      LOG(INFO) << "[import] Succeed to import slot " << slot;
+      break;
+    case kImportFail:
+      if (svr_->slot_import_->Fail(slot) == false) {
+        LOG(ERROR) << "[import] Failed to set slot importing error, maybe slot is wrong"
+                  << ", recieved slot: " << slot
+                  << ", current slot: " << svr_->slot_import_->GetSlot();
+        return Status(Status::NotOK, "Failed to set slot " + std::to_string(slot) + " importing error");
+      }
+      LOG(INFO) << "[import] Failed to import slot " << slot;
+      break;
+    default:
+      return Status(Status::NotOK, "Invalid import state");
+      break;
+  }
+  return Status::OK();
+}
+
+Status Cluster::GetMigrateInfo(int slot, std::vector<std::string> *info) {
+  if (myself_->role_ != kClusterMaster || svr_->IsSlave()) {
```

Comment:
I saw the role and slot check in many places, we can become it to the member function.